tools/bcdb_mock/mw/auth.go

```diff
@@ -0,0 +1,103 @@
+package mw
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zos/pkg/schema"
+	"github.com/threefoldtech/zos/tools/bcdb_mock/pkg/phonebook/types"
+	"github.com/zaibon/httpsig"
+
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+type mongoKeyGetter struct {
+	db *mongo.Database
+}
+
+func (m mongoKeyGetter) GetKey(id string) interface{} {
+	ctx := context.TODO()
+
+	uid, err := strconv.ParseInt(id, 10, 64)
+	if err != nil {
+		return nil
+	}
+
+	f := types.UserFilter{}
+
+	f = f.WithID(schema.ID(uid))
+
+	col := m.db.Collection(types.UserCollection)
+	result := col.FindOne(ctx, f, options.FindOne())
+
+	err = result.Err()
+	if err != nil {
+		return nil
+	}
+
+	user := types.User{}
+	if err = result.Decode(&user); err != nil {
+		return nil
+	}
+
+	pk, err := hex.DecodeString(user.Pubkey)
+	if err != nil {
+		return nil
+	}
+	return pk
+}
+
+// requiredHeaders are the parameters to be used to generated the http signature
+var requiredHeaders = []string{"(created)", "date", "threebot-id"}
+
+func AuthMiddleware(db *mongo.Database, h http.Handler) http.Handler {
+	kg := mongoKeyGetter{db}
+	verifier := httpsig.NewVerifier(kg)
+	verifier.SetRequiredHeaders(requiredHeaders)
+
+	return requireSignature(h, verifier)
+}
+
+func requireSignature(h http.Handler, v *httpsig.Verifier) (
+	out http.Handler) {
+
+	var challengeParams []string
+	if headers := v.RequiredHeaders(); len(headers) > 0 {
+		challengeParams = append(challengeParams,
+			fmt.Sprintf("headers=%q", strings.Join(headers, " ")))
+	}
+
+	challenge := "Signature"
+	if len(challengeParams) > 0 {
+		challenge += fmt.Sprintf(" %s", strings.Join(challengeParams, ", "))
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		err := v.Verify(req)
+		if err != nil {
+			w.Header()["WWW-Authenticate"] = []string{challenge}
```

Comment:
I don't get this actually, does this mean that the user (or the node) has to do 2 trips to execute an action. First one to get Unauthorized + the header to sign, then a request with the correct signature ?

Reply:
Well if you send a request with the proper header right away then you will not get the 401 error.
Here we just notify the client that if he wants to authenticate, it needs to use these header to create the signature.

Reply:
Yes, I understand but if the client will not know what are the required headers unless at least one call is made. right? also does the challenge change with every request?

Reply:
The challenge doesn't not change with every request no. And this needs to be documented. Normal user will always include this during each requests.